Reduce stick-to-air latency in some protocols with manual mixer scheduling

hexfet opened the pull request:
New function allows protocols to schedule the input sampling and mixer calculations.  Provides a means for protocols to reduce input-to-air latency by having inputs sampled and mixers run just before the protocol packet is assembled.  Protocols not using the new API remain unchanged, with input sampling and mixer calculations running every 5ms independent of protocol timing.

Measurements and discussion of design considerations are in[ this thread](https://www.deviationtx.com/forum/7-development/7943-latency-test-for-deviation-vs-ersky-vs-opentx).

Functions added:
 void CLOCK_RunMixer(volatile int *sync): This function generates an interrupt to sample the ADC inputs and run the mixer calculations.  Protocols can use this function to cause updated Channels values to be ready just when they are needed by the protocol.  The sync variable is cleared to 0 when this function is called, then set to 1 when the mixer update is complete.  The time required to complete mixer calculations depends on the complexity of the mixers in the model.  Protocols can check this variable to ensure the mixer calculations are complete when used, and adjust timing to ensure the minimum latency.  See frskyx_cc2500.c for an example.

 void CLOCK_StartMixer(): Starts mixers running on periodic 5ms timer (same as previous behavior).  This is the default and is called every time a new protocol is loaded.

The protocols that are updated to reduce latency by scheduling mixers in the protocol are: crsf, devo, flysky, frsky, frskyx, sbus, sfhss.  The sbus protocol also gets a new option to adjust packet interval.

hexfet commented on src/protocol/frskyx_cc2500.c:
```diff
@@ -755,12 +759,30 @@ static u16 frskyx_cb() {
+          datam_state = FRSKY_DATA4;
+          return 3100 - mixer_runtime;
+      }
       return 3100;
```
Removed.

hexfet commented on src/main.c:
```diff
@@ -199,13 +199,7 @@ void Banner()
 
 void medium_priority_cb()
```
I like this suggestion.  Left it in for reference but if no one objects I'll remove it.

hexfet commented on src/protocol/crsf_uart.c:
```diff
@@ -276,31 +276,52 @@ static void build_rcdata_pkt()
+        return mixer_runtime;
+
+    case ST_DATA2:
+        if (!mixer_sync && mixer_runtime < 2000) mixer_runtime += 50;
```
The algorithm checks if the mixer calculations completed before being used.  If it did not then add 50 microseconds to its next runtime.  The mixer_runtime variable is used in the callback state machine to adjust when the mixer is run.

The forum thread describes the testing - a model with the maximum number of mixes takes about 1500us to run the mixers.  Simple models take much less time. Here mixer_runtime is initialized to 50us.  If that's not enough for the mixers to complete, it is incremented by 50us each packet interval until the mixers do complete.  This keeps the latency at the minimum working value (within 50us or so).  It only takes a few packet intervals at startup time to reach the working value.

howard0su replied:
Then I understand. I can help take a look mixer code to see why it is so slow. can u point me the exact function here?

if mixer_sync is not finish, shall we just return 50 to give us another try in 50us later?

hexfet replied:
A good place to start is MIXER_CalcChannels.

No that would change the protocol timing.  The protocols that dynamically adjust the runtime given to the mixer calculations quickly adjust the runtime up to ensure the mixer calculations are complete.  For a typical 6-channel model the testing I've done shows the mixer runtime is about 450-500 microseconds.  Since the runtime is adjusted in 50 us steps it takes less than a dozen packet intervals to reach a good value - less than 200 ms.

howard0su replied:
got you. I think your logic is basically try to find out the max latency of mixer calculation. Ultimately, we should limit the calculation time of mixer.

hexfet commented on src/target/common/stm32/clock.c:
```diff
@@ -219,46 +219,64 @@ void CLOCK_ClearMsecCallback(int cb)
 
+// Run Mixer one time.  Used by protocols that trigger mixer calc in protocol code
+static volatile int *mixer_sync;
+void CLOCK_RunMixer(volatile int *sync) {
```
Just avoiding making mixer_sync a global since it must be accessed from within the protocols.  But we have plenty of other globals so that might fit better.  I'll consider this change.

hexfet replied:
Using a global for mixer_sync does make the code look cleaner I think.  Pushed the change. Good suggestion.

hexfet commented on src/protocol/sbus_uart.c:
```diff
@@ -84,12 +94,33 @@ static void build_rcdata_pkt()
-    UART_Send(packet, sizeof packet);
-
-    return SBUS_FRAME_PERIOD;
+    if (sbus_period != Model.proto_opts[PROTO_OPTS_PERIOD] * 1000)
```
Well I hadn't noticed SETOPTIONS.  Seems like it opens a synchronization issue.  The options are used within an ISR so updating through SETOPTIONS opens the possibility of using a partially updated value.

howard0su replied:
It is not a problem to atomic read/write. check this:
https://stackoverflow.com/questions/8107108/stm32-programming-tips-and-questions

Question 2: Sharing variables between interrupts and the main loop

People have written entire chapters explaining the answers to this group of questions. Whenever you share a variable between the main loop and an interrupt, you should definitely use the volatile keywords on it. Variables of 32 or fewer bits can be accessed atomically (unless they are misaligned).

If you need to access a larger variable, or two variables at the same time from the main loop, then you will have to disable the clock interrupt while you are accessing the variables. If your interrupt does not require precise timing, this will not be a problem. When you re-enable the interrupt, it will automatically fire if it needs to.

hexfet replied:
The protocol options are 16bit in the Model struct and no guarantee they are aligned (though I guess we could force that).  Seems like that's better suited to a separate PR, especially since most protocols the same technique as here so they could all be cleaned up.

howard0su replied:
sounds good to me. actually, i believe protocol code need some cleanup after we branch out for a new release. I am not happy with the fact all protocol has its own statemachine. We should consolidate the statemachine. Your change on mixer can be also implement in one single place.

hexfet commented on src/protocol/frskyx_cc2500.c:
```diff
@@ -702,6 +702,8 @@ static const u8 telem_test[][17] = {
 #endif
 
 
+static volatile int mixer_sync;
```
Implementation changed.

PhracturedBlue commented:
Overall, I like this approach, though I am concerned what happens if it takes more time to calculate than one iteration of the protocol state machine.  That probably can't happen given the timings you've seen. 
 I'm actually really surprised that it can take 1.2msec to compute the Mixers.  That is a lot longer than I'd expect.

Also note that some Transmitters (like the devo12) have very noisey ADC circuits due to PCB layout, and I believe I ended up using much heavier averaging on them.  I don't know if it affects the overall timing or not though.

I didn't do a really deep dive into the patch, but I'd expect to see a lot of testing, especially for frsky which as I recall was a really problematic protocol from a timing perspective.

PhracturedBlue commented:
sounds good.  I did substantial testing of the ADC when I implemented it originally.  I believe the values are pretty much optimized for Devo 8/10/12/7e (possibly a little conservative for the 8/10).  I know I tried to balance the latency against the uncertainty.  The 12 was an outlier in how noisy it was but I think the CPU used is much faster and the ADC resolves quicker?  I dunno, it was a long time ago.  I don't think we ever even looked at any of the other transmitters since the values we had were good enough.  If any of them implemented proper shielding of the analog wires, the averaging could certainly be reduced though.

phobos- commented:
I really anticipated this feature and am willing to test it. I can test frsky flysky and sfhss . If you modify dsm2/dsmx I can test it as well.

hexfet commented:
Modified DSM protocols for reduced latency, but the changes would not fit in modular builds.  Test build is updated (850d7dd) but I won't be able to test until tomorrow.

hexfet commented:
Reverted commit 459b which changed the mixer_sync implementation from passed pointer to global.  No idea why it fails.  The assembly code is very similar for both implementations, with the difference being the additional indirection.  But with the global implementation the runtime behavior seems to not be consistent with the assembly code in the list file.  For example, selecting an unmodified protocol so the mixer runs on a timer same as master, the comparison of mixer_sync to MIX_TIMER fails and the mixer is never run.  Will continue to look at the global implementation but reverting 459b gives a working test build (f6223b0), which is updated on the Test Builds page.

Except for DSM which seems to be buggy :)  Will debug that this weekend.

hexfet commented:
Fixed the issue with the global implementation - just a regular bug.  Fixed the DSM protocol.  The [test build](https://www.deviationtx.com/downloads-new/category/1504-hexfet-reduce-latency) is updated (8d59ce1).

I've tested sfhss and dsmx.  Will make a post asking for testing on the forum after checking the others.

PhracturedBlue commented:
As I'm writing tests for the mixer now, I realized that this change may break the MUX_DELAY mechanism.  Have you tested it?
If not, please do...I'll have a test for it pretty soon which may help

PhracturedBlue commented:
...on further thought, my tests won't help at all. since the problem is the frequency of calling the mixer which can't be tested in unit tests.  My guess is you'll need to keep track of the # of milliseconds since the last mixer update and use that to compute the delay.

phobos- commented:
I happen to have a MUX_DELAY in my mixer, I use it to have turtle and arm on one switch, it doesn't work without delay in velocidrone. Will test soon and report here.

phobos- commented:
OK, delay is working fine (maybe a little slower (?)), and so is dsmx, flysky afhds 2a and frsky x eu. I checked for packetloss in betaflight and it looks fine, nothing out of ordinary.
I didnt notice any side effects so far. I feel maybe 1% more confident, but it might be placebo effect.
Will test it some more.

PhracturedBlue commented:
Sorry, that isn't really what I meant.  I'm sure the delay will do something, but I expect it to be too slow and to potentially vary per protocol.  There is a specific rate we expect to see.  I believe the code says that the movement should be 60 degrees in channel-value / 20 seconds.

It certainly should not be arbitrarily defined.

hexfet commented:
It should only be different for the protocols that have optimized latency.  For all the rest the code is effectively unchanged (mixer run triggered every MEDIUM_PRIORITY_MSEC).  I've coded a fix but haven't tested yet.  Finally had nice weekend weather...

Thanks for the protocol testing.

hexfet commented:
The [test build is updated](https://www.deviationtx.com/downloads-new/category/1511-hexfet-reduce-latency) (ca5de44).  Tested the delay mux (and apply limits) fix on T8+ and devo10.  Timing is the same when compared with master for both unmodified and reduced latency protocols within eyeball accuracy.  Timed with stopwatch, and compared transmitters running master and reduce_latency side-by-side.

brycedjohnson commented:
Let me know and I can run latency test setup for end to end latency for frsky_x and CSFR that I already have data for.

hexfet commented:
@brycedjohnson Thanks that would be great.  The results should be similar to the last test build.

Although the mixer code seems to be running faster.  Seeing 20us for minimum mixers model, and 740us with maximum mixers model.  Previously I'd seen values over 1ms for max mixers.  Using a different test method so that may be the difference.  Probably not a significant enough difference to show up in your tests.
